### ISSUE TITLE
Serial output not readable

### ISSUE THREAD

**Issue by kitzler-walli** [tirolerstefan/kaifa#1]
Currently I try to get the data from my MA309 with your recommended hardware but it seems like the output coming from the serial port is almost garbage - any hints for me? 

Output:
````
?~e?6}b?T&??????OR"?շ??_l?㔬?9??m[2?4zC2??F???Y???]?
??gXqm?
       ?2u???̦?ʾ?
?:?e???М?0Н?8 ?BFm??h??hS?gKFMeP?j??? F????? 
                   <&^??h_?(/?`??O??ۧ???OY?_r(H?-???}s??#$?`A?z??jB???c?!??l?
````

**Comment by tirolerstefan** [comment]
Hi, what's your energy supplier? Do you have a spec about the protocol?
I would check the serial configuration, then.

**Comment by kitzler-walli** [comment]
just one question in the meantime - how to know which wire is plus/minus between JOYKK and 6P6C?

**Comment by tirolerstefan** [comment]
Just try each wiring and check the output.

**Comment by kitzler-walli** [comment]
Tried it both ways but the output was cryptic both times.

**Comment by kitzler-walli** [comment]
The key I received from EVN is 32 characters long - and seems to be hex.
Converting it to hex would result in a 64 character long string. Which one should I use?
Any difference if it's lower or upper case? (the one I received was uppercase)

**Comment by tirolerstefan** [comment]
Ok, this is an example how to decrpyt your data, you should try this, first, with the above stream:
https://www.gurux.fi/node/18100

The example in this link is exactly the same as in your EVN PDF. But as Mikko says, the example data seems to be not valid. Anyway, we have real data from your logging, above.

So, please try the following:
1. Install the gurux libs (https://github.com/Gurux/Gurux.DLMS.Python)
2. In your python console (as far I saw, it is python2), try the following:

```
bb = GXByteBuffer(stream_from_meter)
cl = GXDLMSSecureClient()
data = GXReplyData()
notify = GXReplyData()
cl.interfaceType = InterfaceType.PDU
cl.ciphering.security = Security.ENCRYPTION
cl.ciphering.blockCipherKey = GXCommon.hexToBytes(key_hex)
plaintext = cl.getData(bb, data, notify)
print(notify.value)
```

... whereas "Value coming form the meter" is:
```
stream_from_meter=b"\x1f\x95s\xfb\xdc\x91d\xd1\xcf\td\x0c$\xed)=V\xb8w\x02\x15~'U\x8c\xc4\x11\xfa\xae\x89\x0f3\xcbW\xb9j\xe2\xa8\x11wRn\xd3^\x96\xe4\xab\xf8F\x88$n?\xc2\xdcW\x0bj\xef\xd2\xda\x88V\xdb\xbc\xea\xd8\x97\x82\xc9e\xfc\xf4X\x89%\x8f\x8cY,\xd3v\xee\xa5D\x16\xfe5\xb8q\xc8I\x1ev\n\xfd\xf9\x8a\xbd\xcb?\x1cm\x96qh\xef\xb5c\xa2\xb60%Gg\xb2\xcf\x19\xbe\x00\xf8\x82\x86^_\xdf\xfd\x81\xfb\xac<5\xc0\x07\x81 \xcfgt\x0b\x8d\x85\xdc{\x98C\xdf\x94M\xc8:jL!\xfa\xeb\xcb\xed-t\xdd\x15\x8d\x16x{y\x9d`&\x07\xce\xc4\x8eD\xd7T^S\xe9\xdew\xa2\x8b\x1a&\x81\xe1\xf3\xe7/\xb3U\xb7\x1d\xb0pu\xd8\x81Pc\xf1\x10\x9e\t0gYba\x9a\xc4\x86Whkat\xdc\x00\xca\xd2Wp\x93\x90kw\xb5\xef\xa6>{+O\xa9p5\x19n\xb3"
```

If I concatenated it incorrectly, here are the hey byte strings of the data of both telegrams:
```
data_frame1=b'1f9573fbdc9164d1cf09640c24ed293d56b87702157e27558cc411faae890f33cb57b96ae2a81177526ed35e96e4abf84688246e3fc2dc570b6aefd2da8856dbbcead89782c965fcf45889258f8c592cd376eea54416fe35b871c8491e760afdf98abdcb3f1c6d967168efb563a2b630254767b2cf19be00f882865e5fdffd81fbac3c35c0078120cf67740b8d85dc7b9843df944dc83a6a4c21faebcbed2d74dd158d16787b799d602607cec48e44d7545e53e9de77a28b1a2681e1f3e72fb355b71db07075d8815063f1109e0930675962619ac48657686b6174dc00cad257709390'

data_frame2=b'6b77b5efa63e7b2b4fa97035196eb3'

data_frames=b'1f9573fbdc9164d1cf09640c24ed293d56b87702157e27558cc411faae890f33cb57b96ae2a81177526ed35e96e4abf84688246e3fc2dc570b6aefd2da8856dbbcead89782c965fcf45889258f8c592cd376eea54416fe35b871c8491e760afdf98abdcb3f1c6d967168efb563a2b630254767b2cf19be00f882865e5fdffd81fbac3c35c0078120cf67740b8d85dc7b9843df944dc83a6a4c21faebcbed2d74dd158d16787b799d602607cec48e44d7545e53e9de77a28b1a2681e1f3e72fb355b71db07075d8815063f1109e0930675962619ac48657686b6174dc00cad2577093906b77b5efa63e7b2b4fa97035196eb3'

stream_from_meter=binascii.unhexlify(data_frames)
```

Let me know whether it worked.

**Comment by kitzler-walli** [comment]
Here's the output of the program you supplied - seems like there is problem with the sample data:

```
Traceback (most recent call last):
  File "test.py", line 17, in <module>
    bb = GXByteBuffer(stream_from_meter)
  File "/usr/local/lib/python2.7/dist-packages/gurux_dlms/GXByteBuffer.py", line 65, in __init__
    self.set(value)
  File "/usr/local/lib/python2.7/dist-packages/gurux_dlms/GXByteBuffer.py", line 444, in set
    value = value.encode()
UnicodeDecodeError: 'ascii' codec can't decode byte 0x95 in position 1: ordinal not in range(128)
```

**Comment by dbeinder** [comment]
@tirolerstefan The COSEM length field is 0x81 +1 byte in the EVN telegram, it needs to be decoded like this: https://github.com/dbeinder/KsmWestKaifaReader/blob/master/kaifa.py#L142-L150
Because of this, the encrypted frame already starts at 0x24 0x1F 0x95
@kitzler-walli try adding `\x24` at the start of `stream_from_meter`

**Comment by kitzler-walli** [comment]
@dbeinder I tried as you said but the error persisted as byte 0x95 still can't be decoded.
In the meantime I tried to use your solution but that resulted in this error:
```
Decrypted:
0f8006e3620c07e50a110710350500ff88800223090c07e50a110710350500ff888009060100010800ff0600049da802020f00161e09060100020800ff060000355f02020f00161e09060100010700ff060000006902020f00161b09060100020700ff060000000002020f00161b09060100200700ff12091f02020fff162309060100340700ff12092b02020fff162309060100480700ff12092902020fff1623090601001f0700ff12002602020ffe162109060100330700ff1200df02020ffe162109060100470700ff12005d02020ffe1621090601000d0700ff12015502020ffd16ff090c313831323130313835333734
Unsupported OBIS code length 12
Traceback (most recent call last):
  File "/home/pi/KsmWestKaifaReader/kaifa.py", line 211, in <module>
    timestamp = obis[Obis.Timestamp]
KeyError: b'\x00\x00\x01\x00\x00\xff'
```

**Comment by tirolerstefan** [comment]
@kitzler-walli - I think you should try @dbeinder 's script - if this is already supporting EVN telegrams!
In the meantime, I'll try to check the hint regarding length field - thanks for that, David.

**Comment by kitzler-walli** [comment]
Thank you for the update - but the result stays the same:
```
2021-10-17 17:47:29,442 [DEBUG]:  pos: 168 | 142 | 168
2021-10-17 17:47:44,697 [DEBUG]:  pos: 0 | 256 | 282
2021-10-17 17:47:44,699 [DEBUG]:  b'68fafa6853ff000167db084b464d65509b6a9e81f82000066865cc35b14d815badf2d0b5105a85819557c3501bf932f5a619b43bf7e90a65a3ff69ae9c147120c796df40fad277563fa390bdde896db36bf0a179fe04e4faba8d262bd1d28628e3fa1bcedf99d11c1bb62a2cb826c05ff8837bf641cfa8fd8550f4e01454e5c257a1f639a72af117de171a8e0f4c8cedd0d7b3e80b98ca23555d7035853a4a926b2112a1fcfac3c176b99a671657fdef6a4de9fb8af3fe90c62bbe36929fe1a229a43f6aff4bca00445279ddde39456efcc05d2c7834f6e4da440e045f65c18c18d978e6f7ec4c42a6cc20163a77dcabd5a985ced9095166845fd8d8f49cbf166814146853ff110167bc8da24d15b936e0b8dbd21abacfb7a61668fafa6853ff000167db084b464d65509b6a9e81f82000066866610fe9c11dc0d1917f3d1bd484959f78893ef435dda69d5c2c686b405af795f756d3ef6ceb7a56e3ce5963db3e66084eb059e7418b29a1b69850401720745deb5bccfc2c9ef5322920613b72883fa8b55b515f3640211a04409ed187186130fdc8d8e79087b2a8365a474a303995cf4a805e88b8da01d9831ecfd0147e53dc86c289e52363070bb5fd48c30cce169892f5329b43474560b95805dd5236054819b7ea8abf6979565e05afb8038c555508441e006ffa3f437036b65c6faee1664e4b864ea05fbf45dc58c39853975a23c3491cf32efce254bc15df52f4e6d87c8c1a48dede4a166814146853ff110167cde3b3e38b948003363400ccf5fde3be1668fafa6853ff000167db084b464d65509b6a9e81f8200006686799caa2f8f285fe20beaac067981a970b5282ec1a88b81ad8bc02d934f4de41e689dc80b8f4d755269b01ca240341104628d6510697f6215eacf933e52c409477e069fe18fc576dda506e09a79c19854534f2eced727940460ffad9408cd26cb07fa21e5f829251d3edfc445bd4ed3c0b6226acb0596c9b36f7934c5b5544f3c4da60b58b75629986aa0377c091dffb314fafcea0604817511152ac5f5ad56e8807f38889b593128172262afcc0e76595e33f389865ba581693a06611de3e32d43a0f5caf42aa7fcf3eeef25061421b13d8e5f7486e2c66af5f2551dc3f9f2e195a2d22524d166814146853ff110167c75aefeb09a8c3e4c7e9d7a8f564e68c1668fafa6853ff000167db084b464d65509b6a9e81f82000066868a94a8b98d85b5ce1b7ca30865b70828b0f5c645007056be99b49e447c644b755f2c8a18232978f3882b97bb27fd7f7d2ec8efe0032fd79d947741fc5bb23e61c607adc29bd51604fd46a25a4ca01528aa9e70a'
2021-10-17 17:47:44,700 [DEBUG]:  [b'h\xfa\xfah', b'S\xff\x00\x01g\xdb\x08KFMeP\x9bj\x9e\x81\xf8 \x00\x06he\xcc5\xb1M\x81[\xad\xf2\xd0\xb5\x10Z\x85\x81\x95W\xc3P\x1b\xf92\xf5\xa6\x19\xb4;\xf7\xe9\ne\xa3\xffi\xae\x9c\x14q \xc7\x96\xdf@\xfa\xd2wV?\xa3\x90\xbd\xde\x89m\xb3k\xf0\xa1y\xfe\x04\xe4\xfa\xba\x8d&+\xd1\xd2\x86(\xe3\xfa\x1b\xce\xdf\x99\xd1\x1c\x1b\xb6*,\xb8&\xc0_\xf8\x83{\xf6A\xcf\xa8\xfd\x85P\xf4\xe0\x14T\xe5\xc2W\xa1\xf69\xa7*\xf1\x17\xde\x17\x1a\x8e\x0fL\x8c\xed\xd0\xd7\xb3\xe8\x0b\x98\xca#U]p5\x85:J\x92k!\x12\xa1\xfc\xfa\xc3\xc1v\xb9\x9ag\x16W\xfd\xefjM\xe9\xfb\x8a\xf3\xfe\x90\xc6+\xbe6\x92\x9f\xe1\xa2)\xa4?j\xffK\xca\x00DRy\xdd\xde9En\xfc\xc0],x4\xf6\xe4\xdaD\x0e\x04_e\xc1\x8c\x18\xd9x\xe6\xf7\xecLB\xa6\xcc \x16:w\xdc\xab\xd5\xa9\x85\xce\xd9\tQf\x84_\xd8\xd8\xf4\x9c\xbf\x16', b'h\x14\x14h', b'S\xff\x11\x01g\xbc\x8d\xa2M\x15\xb96\xe0\xb8\xdb\xd2\x1a\xba\xcf\xb7\xa6\x16', b'h\xfa\xfah', b'S\xff\x00\x01g\xdb\x08KFMeP\x9bj\x9e\x81\xf8 \x00\x06hfa\x0f\xe9\xc1\x1d\xc0\xd1\x91\x7f=\x1b\xd4\x84\x95\x9fx\x89>\xf45\xdd\xa6\x9d\\,hk@Z\xf7\x95\xf7V\xd3\xefl\xebzV\xe3\xceYc\xdb>f\x08N\xb0Y\xe7A\x8b)\xa1\xb6\x98P@\x17 t]\xeb[\xcc\xfc,\x9e\xf52) a;r\x88?\xa8\xb5[Q_6@!\x1a\x04@\x9e\xd1\x87\x18a0\xfd\xc8\xd8\xe7\x90\x87\xb2\xa86ZGJ09\x95\xcfJ\x80^\x88\xb8\xda\x01\xd9\x83\x1e\xcf\xd0\x14~S\xdc\x86\xc2\x89\xe5#c\x07\x0b\xb5\xfdH\xc3\x0c\xce\x16\x98\x92\xf52\x9bCGE`\xb9X\x05\xddR6\x05H\x19\xb7\xea\x8a\xbfiyV^\x05\xaf\xb8\x03\x8cUU\x08D\x1e\x00o\xfa?Cp6\xb6\\o\xae\xe1fNK\x86N\xa0_\xbfE\xdcX\xc3\x98S\x97Z#\xc3I\x1c\xf3.\xfc\xe2T\xbc\x15\xdfR\xf4\xe6\xd8|\x8c\x1aH\xde\xdeJ\x16', b'h\x14\x14h', b'S\xff\x11\x01g\xcd\xe3\xb3\xe3\x8b\x94\x80\x0364\x00\xcc\xf5\xfd\xe3\xbe\x16', b'h\xfa\xfah', b'S\xff\x00\x01g\xdb\x08KFMeP\x9bj\x9e\x81\xf8 \x00\x06hg\x99\xca\xa2\xf8\xf2\x85\xfe \xbe\xaa\xc0g\x98\x1a\x97\x0bR\x82\xec\x1a\x88\xb8\x1a\xd8\xbc\x02\xd94\xf4\xdeA\xe6\x89\xdc\x80\xb8\xf4\xd7U&\x9b\x01\xca$\x03A\x10F(\xd6Q\x06\x97\xf6!^\xac\xf93\xe5,@\x94w\xe0i\xfe\x18\xfcWm\xdaPn\t\xa7\x9c\x19\x85E4\xf2\xec\xedry@F\x0f\xfa\xd9@\x8c\xd2l\xb0\x7f\xa2\x1e_\x82\x92Q\xd3\xed\xfcD[\xd4\xed<\x0bb&\xac\xb0Yl\x9b6\xf7\x93L[UD\xf3\xc4\xda`\xb5\x8bub\x99\x86\xaa\x03w\xc0\x91\xdf\xfb1O\xaf\xce\xa0`H\x17Q\x11R\xac_Z\xd5n\x88\x07\xf3\x88\x89\xb5\x93\x12\x81r&*\xfc\xc0\xe7e\x95\xe3?8\x98e\xbaX\x16\x93\xa0f\x11\xde>2\xd4:\x0f\\\xafB\xaa\x7f\xcf>\xee\xf2PaB\x1b\x13\xd8\xe5\xf7Hn,f\xaf_%Q\xdc?\x9f.\x19Z-"RM\x16', b'h\x14\x14h', b'S\xff\x11\x01g\xc7Z\xef\xeb\t\xa8\xc3\xe4\xc7\xe9\xd7\xa8\xf5d\xe6\x8c\x16', b'h\xfa\xfah', b'S\xff\x00\x01g\xdb\x08KFMeP\x9bj\x9e\x81\xf8 \x00\x06hh\xa9J\x8b\x98\xd8[\\\xe1\xb7\xca0\x86[p\x82\x8b\x0f\\dP\x07\x05k\xe9\x9bI\xe4G\xc6D\xb7U\xf2\xc8\xa1\x822\x97\x8f8\x82\xb9{\xb2\x7f\xd7\xf7\xd2\xec\x8e\xfe\x002\xfdy\xd9Gt\x1f\xc5\xbb#\xe6\x1c`z\xdc)\xbdQ`O\xd4j%\xa4\xca\x01R\x8a\xa9\xe7\n']
2021-10-17 17:47:44,701 [DEBUG]:  TELEGRAM1:
b'68fafa6853ff000167db084b464d65509b6a9e81f82000066865cc35b14d815badf2d0b5105a85819557c3501bf932f5a619b43bf7e90a65a3ff69ae9c147120c796df40fad277563fa390bdde896db36bf0a179fe04e4faba8d262bd1d28628e3fa1bcedf99d11c1bb62a2cb826c05ff8837bf641cfa8fd8550f4e01454e5c257a1f639a72af117de171a8e0f4c8cedd0d7b3e80b98ca23555d7035853a4a926b2112a1fcfac3c176b99a671657fdef6a4de9fb8af3fe90c62bbe36929fe1a229a43f6aff4bca00445279ddde39456efcc05d2c7834f6e4da440e045f65c18c18d978e6f7ec4c42a6cc20163a77dcabd5a985ced9095166845fd8d8f49cbf16'

2021-10-17 17:47:44,703 [DEBUG]:  TELEGRAM2:
b'6814146853ff110167bc8da24d15b936e0b8dbd21abacfb7a616'

2021-10-17 17:47:44,704 [DEBUG]:  Decrypt: FRAME1:
b'68fafa6853ff000167db084b464d65509b6a9e81f82000066865cc35b14d815badf2d0b5105a85819557c3501bf932f5a619b43bf7e90a65a3ff69ae9c147120c796df40fad277563fa390bdde896db36bf0a179fe04e4faba8d262bd1d28628e3fa1bcedf99d11c1bb62a2cb826c05ff8837bf641cfa8fd8550f4e01454e5c257a1f639a72af117de171a8e0f4c8cedd0d7b3e80b98ca23555d7035853a4a926b2112a1fcfac3c176b99a671657fdef6a4de9fb8af3fe90c62bbe36929fe1a229a43f6aff4bca00445279ddde39456efcc05d2c7834f6e4da440e045f65c18c18d978e6f7ec4c42a6cc20163a77dcabd5a985ced9095166845fd8d8f49cbf16'
2021-10-17 17:47:44,705 [DEBUG]:  Decrypt: FRAME2:
b'6814146853ff110167bc8da24d15b936e0b8dbd21abacfb7a616'
2021-10-17 17:47:44,706 [DEBUG]:  SYSTITLE: b'4b464d65509b6a9e'
2021-10-17 17:47:44,707 [DEBUG]:  IC: b'066865cc' / 107505100
2021-10-17 17:47:44,708 [DEBUG]:  IV: b'4b464d65509b6a9e066865cc'
2021-10-17 17:47:44,710 [DEBUG]:  DATA FRAME1
b'cc35b14d815badf2d0b5105a85819557c3501bf932f5a619b43bf7e90a65a3ff69ae9c147120c796df40fad277563fa390bdde896db36bf0a179fe04e4faba8d262bd1d28628e3fa1bcedf99d11c1bb62a2cb826c05ff8837bf641cfa8fd8550f4e01454e5c257a1f639a72af117de171a8e0f4c8cedd0d7b3e80b98ca23555d7035853a4a926b2112a1fcfac3c176b99a671657fdef6a4de9fb8af3fe90c62bbe36929fe1a229a43f6aff4bca00445279ddde39456efcc05d2c7834f6e4da440e045f65c18c18d978e6f7ec4c42a6cc20163a77dcabd5a985ced9095166845fd8d8f49c'
2021-10-17 17:47:44,711 [DEBUG]:  DATA FRAME1
b'bc8da24d15b936e0b8dbd21abacfb7'
2021-10-17 17:47:44,714 [DEBUG]:  b'd41239f6047243b0e6f37d4d986f85d7007bb4d72238e77dd4ed7abd7d23957a8ae7b65f2e3e45fca0903eef10f11bab49d745ca018ee3869b102a37e7383ed18557df45eee2083e0e8743411448371c261c756da95b14698af0c48f72610bb2eb7b7c8d6472685ef6a1ff23984419f2af68bcb356c87a0654d55d74c449042647f71f4a0f5d8d85616ca64a657128bdf081c08058925c49aa1fd42bcb52ade4fc6883d1967fc80d87ab26bb68f93efb034fee857eee02d53f740c6b5353ef6ae6603d94b6e485e0d85764b6a5111c9db93a5748e6caee2358eff29b37d37fb13701fa79b778e250ff7478f76c139299822fd7'
2021-10-17 17:47:44,715 [DEBUG]:  [b'd41239f6047243b0e6f37d4d986f85d7007bb4d72238e77dd4ed7abd7d23957a8ae7b65f2e3e45fca0903eef10f11bab49d745ca018ee3869b102a37e7383ed18557df45eee2083e0e8743411448371c261c756da95b14698af0c48f72610bb2eb7b7c8d6472685ef6a1ff23984419f2af68bcb356c87a0654d55d74c449042647f71f4a0f5d8d85616ca64a657128bdf081c08058925c49aa1fd42bcb52ade4fc6883d1967fc80d87ab26bb68f93efb034fee857eee02d53f740c6b5353ef6ae6603d94b6e485e0d85764b6a5111c9db93a5748e6caee2358eff29b37d37fb13701fa79b778e250ff7478f76c139299822fd7']
2021-10-17 17:47:44,716 [DEBUG]:  b'd41239f6047243b0e6f37d4d986f85d7007bb4d72238e77dd4ed7abd7d23957a8ae7b65f2e3e45fca0903eef10f11bab49d745ca018ee3869b102a37e7383ed18557df45eee2083e0e8743411448371c261c756da95b14698af0c48f72610bb2eb7b7c8d6472685ef6a1ff23984419f2af68bcb356c87a0654d55d74c449042647f71f4a0f5d8d85616ca64a657128bdf081c08058925c49aa1fd42bcb52ade4fc6883d1967fc80d87ab26bb68f93efb034fee857eee02d53f740c6b5353ef6ae6603d94b6e485e0d85764b6a5111c9db93a5748e6caee2358eff29b37d37fb13701fa79b778e250ff7478f76c139299822fd7'
2021-10-17 17:47:44,717 [INFO]:  1.8.0: 0
2021-10-17 17:47:44,718 [INFO]:  2.8.0: 0
2021-10-17 17:47:54,374 [DEBUG]:  pos: 173 | 147 | 173
2021-10-17 17:47:54,568 [DEBUG]:  pos: 0 | 256 | 282
2021-10-17 17:47:54,571 [DEBUG]:  b'68fafa6853ff000167db084b464d65509b6a9e81f820000668690f3d726b0612ad553264b159a7d1dae249636b21b1c43a7622cbc288a5977d604bae694d9912c8c0ae2ccdc2c519f5a45944e0657765982d8792e76ec464a3f75df77ad5be7d4d1cf258614cd065931e8fcecf90ab1ad047ea8192d0c2c519b3599212ebf8661c8c06c9f9a2436e1ea407e10b7b21ca414049e3aa2846c1bda1911bfa539be8b6800d965563c856fa6d8f46f0e78b0d50dbb9f644f7c9174b1dae8fa234543e488cedb8f1fc50e36bf2ed7ec32366b0ed520771636c6d1e2cecf241d4ccf4869926c6be5a32b597212fc9238e688a6a875646966da6a4f9276a4d1a6e91ac166814146853ff110167fad7195dd0abdeb114b573433a9eb1241668fafa6853ff000167db084b464d65509b6a9e81f8200006686a6a1b239849ef970a4b50a46b502f654c9c65cd4358de6c80c6eed36fdf7cd0ce4a97f9ca65c28a464b4cb8cb3184758eeb0a'
2021-10-17 17:47:54,573 [DEBUG]:  [b'h\xfa\xfah', b'S\xff\x00\x01g\xdb\x08KFMeP\x9bj\x9e\x81\xf8 \x00\x06hi\x0f=rk\x06\x12\xadU2d\xb1Y\xa7\xd1\xda\xe2Ick!\xb1\xc4:v"\xcb\xc2\x88\xa5\x97}`K\xaeiM\x99\x12\xc8\xc0\xae,\xcd\xc2\xc5\x19\xf5\xa4YD\xe0ewe\x98-\x87\x92\xe7n\xc4d\xa3\xf7]\xf7z\xd5\xbe}M\x1c\xf2XaL\xd0e\x93\x1e\x8f\xce\xcf\x90\xab\x1a\xd0G\xea\x81\x92\xd0\xc2\xc5\x19\xb3Y\x92\x12\xeb\xf8f\x1c\x8c\x06\xc9\xf9\xa2Cn\x1e\xa4\x07\xe1\x0b{!\xcaA@I\xe3\xaa(F\xc1\xbd\xa1\x91\x1b\xfaS\x9b\xe8\xb6\x80\r\x96Uc\xc8V\xfam\x8fF\xf0\xe7\x8b\rP\xdb\xb9\xf6D\xf7\xc9\x17K\x1d\xae\x8f\xa24T>H\x8c\xed\xb8\xf1\xfcP\xe3k\xf2\xed~\xc3#f\xb0\xedR\x07qclm\x1e,\xec\xf2A\xd4\xcc\xf4\x86\x99&\xc6\xbeZ2\xb5\x97!/\xc9#\x8eh\x8aj\x87VF\x96m\xa6\xa4\xf9\'jM\x1an\x91\xac\x16', b'h\x14\x14h', b'S\xff\x11\x01g\xfa\xd7\x19]\xd0\xab\xde\xb1\x14\xb5sC:\x9e\xb1$\x16', b'h\xfa\xfah', b'S\xff\x00\x01g\xdb\x08KFMeP\x9bj\x9e\x81\xf8 \x00\x06hjj\x1b#\x98I\xef\x97\nKP\xa4kP/eL\x9ce\xcdCX\xdel\x80\xc6\xee\xd3o\xdf|\xd0\xceJ\x97\xf9\xcae\xc2\x8aFKL\xb8\xcb1\x84u\x8e\xeb\n']
2021-10-17 17:47:54,574 [DEBUG]:  TELEGRAM1:
b'68fafa6853ff000167db084b464d65509b6a9e81f820000668690f3d726b0612ad553264b159a7d1dae249636b21b1c43a7622cbc288a5977d604bae694d9912c8c0ae2ccdc2c519f5a45944e0657765982d8792e76ec464a3f75df77ad5be7d4d1cf258614cd065931e8fcecf90ab1ad047ea8192d0c2c519b3599212ebf8661c8c06c9f9a2436e1ea407e10b7b21ca414049e3aa2846c1bda1911bfa539be8b6800d965563c856fa6d8f46f0e78b0d50dbb9f644f7c9174b1dae8fa234543e488cedb8f1fc50e36bf2ed7ec32366b0ed520771636c6d1e2cecf241d4ccf4869926c6be5a32b597212fc9238e688a6a875646966da6a4f9276a4d1a6e91ac16'

2021-10-17 17:47:54,575 [DEBUG]:  TELEGRAM2:
b'6814146853ff110167fad7195dd0abdeb114b573433a9eb12416'

2021-10-17 17:47:54,576 [DEBUG]:  Decrypt: FRAME1:
b'68fafa6853ff000167db084b464d65509b6a9e81f820000668690f3d726b0612ad553264b159a7d1dae249636b21b1c43a7622cbc288a5977d604bae694d9912c8c0ae2ccdc2c519f5a45944e0657765982d8792e76ec464a3f75df77ad5be7d4d1cf258614cd065931e8fcecf90ab1ad047ea8192d0c2c519b3599212ebf8661c8c06c9f9a2436e1ea407e10b7b21ca414049e3aa2846c1bda1911bfa539be8b6800d965563c856fa6d8f46f0e78b0d50dbb9f644f7c9174b1dae8fa234543e488cedb8f1fc50e36bf2ed7ec32366b0ed520771636c6d1e2cecf241d4ccf4869926c6be5a32b597212fc9238e688a6a875646966da6a4f9276a4d1a6e91ac16'
2021-10-17 17:47:54,578 [DEBUG]:  Decrypt: FRAME2:
b'6814146853ff110167fad7195dd0abdeb114b573433a9eb12416'
2021-10-17 17:47:54,578 [DEBUG]:  SYSTITLE: b'4b464d65509b6a9e'
2021-10-17 17:47:54,579 [DEBUG]:  IC: b'0668690f' / 107505935
2021-10-17 17:47:54,580 [DEBUG]:  IV: b'4b464d65509b6a9e0668690f'
2021-10-17 17:47:54,581 [DEBUG]:  DATA FRAME1
b'0f3d726b0612ad553264b159a7d1dae249636b21b1c43a7622cbc288a5977d604bae694d9912c8c0ae2ccdc2c519f5a45944e0657765982d8792e76ec464a3f75df77ad5be7d4d1cf258614cd065931e8fcecf90ab1ad047ea8192d0c2c519b3599212ebf8661c8c06c9f9a2436e1ea407e10b7b21ca414049e3aa2846c1bda1911bfa539be8b6800d965563c856fa6d8f46f0e78b0d50dbb9f644f7c9174b1dae8fa234543e488cedb8f1fc50e36bf2ed7ec32366b0ed520771636c6d1e2cecf241d4ccf4869926c6be5a32b597212fc9238e688a6a875646966da6a4f9276a4d1a6e91'
2021-10-17 17:47:54,582 [DEBUG]:  DATA FRAME1
b'fad7195dd0abdeb114b573433a9eb1'
2021-10-17 17:47:54,585 [DEBUG]:  b'84ae27d9517476620177ba28761a7eeafbaf57ffea5120fb79afb3c97ad34347a79a06e6b2e52f0970ebdd16853a17573555659b5705d6ba193580357410d1d1465981a5a93378407f3134e2a39c1148baf5cc230a1a238b313cf326c969e0705aa08faa93b956ffa7ac512e2d925082a90b7b6ff62eee59a05a5bc7ffd1d65c858a4b0fddc557b2e7acd021c5621642a2615a70522e2e1c231c19e41e38377e056b9a9addbe2878b655edb2d6330b5b747e4a223ac180dce3a6fa30b3dc18c36c2fb5673d16e69dd3db3960d44c91b385b77271199a408417411befa471d3485e8843c536cfaeb3a03a07c757449e39333683'
2021-10-17 17:47:54,587 [DEBUG]:  [b'84ae27d9517476620177ba28761a7eeafbaf57ffea5120fb79afb3c97ad34347a79a06e6b2e52f0970ebdd16853a17573555659b5705d6ba193580357410d1d1465981a5a93378407f3134e2a39c1148baf5cc230a1a238b313cf326c969e0705aa08faa93b956ffa7ac512e2d925082a90b7b6ff62eee59a05a5bc7ffd1d65c858a4b0fddc557b2e7acd021c5621642a2615a70522e2e1c231c19e41e38377e056b9a9addbe2878b655edb2d6330b5b747e4a223ac180dce3a6fa30b3dc18c36c2fb5673d16e69dd3db3960d44c91b385b77271199a408417411befa471d3485e8843c536cfaeb3a03a07c757449e39333683']
2021-10-17 17:47:54,588 [DEBUG]:  b'84ae27d9517476620177ba28761a7eeafbaf57ffea5120fb79afb3c97ad34347a79a06e6b2e52f0970ebdd16853a17573555659b5705d6ba193580357410d1d1465981a5a93378407f3134e2a39c1148baf5cc230a1a238b313cf326c969e0705aa08faa93b956ffa7ac512e2d925082a90b7b6ff62eee59a05a5bc7ffd1d65c858a4b0fddc557b2e7acd021c5621642a2615a70522e2e1c231c19e41e38377e056b9a9addbe2878b655edb2d6330b5b747e4a223ac180dce3a6fa30b3dc18c36c2fb5673d16e69dd3db3960d44c91b385b77271199a408417411befa471d3485e8843c536cfaeb3a03a07c757449e39333683'
2021-10-17 17:47:54,589 [INFO]:  1.8.0: 0
2021-10-17 17:47:54,590 [INFO]:  2.8.0: 0
2021-10-17 17:47:55,132 [DEBUG]:  pos: -1 | -1 | -1
2021-10-17 17:48:04,803 [DEBUG]:  pos: 206 | 180 | 206
```

**Comment by tirolerstefan** [comment]
fixed invocation counter position, change pushed...

**Comment by dbeinder** [comment]
My code only works with KsmWest format, but this shows that decryption works without changes between suite 0 and 1.
You can paste the decrypted data (https://github.com/tirolerstefan/kaifa/issues/1#issuecomment-945139934) in here https://www.gurux.fi/GuruxDLMSTranslator to decode the structure.
The decrypted EVN format is "flatter", the KsmWest format uses a structure entry around each (id,value) pair.
I'm not planning to make my script more universal, but you are welcome to adapt it / copy anything from it.

**Comment by tirolerstefan** [comment]
sorry, invocation counter length wrong ... fixed and pushed that, too :-)

**Comment by kitzler-walli** [comment]
@tirolerstefan that fixed it - thanks a lot
```
2021-10-17 18:36:59,599 [INFO]:  1.8.0: 302.632
2021-10-17 18:36:59,600 [INFO]:  2.8.0: 13.663
```

**Comment by kitzler-walli** [comment]
I'll try to add the extra data that dbeinder extracts in his script to your script if that's ok for you.

**Comment by tirolerstefan** [comment]
of course, no problem. you can use branch "evn" for this.

**Comment by kitzler-walli** [comment]
already tested your supplier change with my additional values - works like charm.
Once again - thank you.

**Comment by tirolerstefan** [comment]
I merged your PR #2, but changed again in PR #4, because it was not working for TiNetz telegrams.
Now, the master works with TiNetz - it would be great if you could test this with your EVN meter, again.
Afterwards, I could finalize this and create a debian release package.
